### PR TITLE
basic service discovery

### PR DESCRIPTION
**Pull request by waynz0r**
## Description

basic service discovery support

## Type of Change

- [ ] Bug Fix
- [x] New Feature
- [ ] Breaking Change
- [ ] Refactor
- [ ] Documentation
- [ ] Other (please describe)

### CODE REVIEW

File: device_driver.c [hunk 1]
```diff
@@ -333,6 +388,25 @@ static int parse_command(const char *data)
...
             }
         }
+        else if (strcmp("load_sd_info", command) == 0)
+        {
+            const char *code = json_object_get_string(root, "code");
+            char *decoded = kzalloc(strlen(code) * 2, GFP_KERNEL);
+            int length = base64_decode(decoded, strlen(code) * 2, code, strlen(code));
+            if (length < 0)
+            {
+                pr_crit("base64_decode failed");
```
Comment (baluchicken) [inline device_driver.c:398]:
```suggestion
                pr_crit("nasp: base64_decode failed");
```

File: socket.c [hunk 2]
```diff
@@ -932,6 +933,7 @@ static int cache_and_validate_cert(nasp_socket *sc, char *key)
 	// Cert found in the cache use that
 	else
 	{
+		pr_err("nasp: found cert for id[%s]", key);
```
Comment (baluchicken) [inline socket.c:936]:
```suggestion
		pr_info("nasp: found cert for id[%s]", key);
```

File: socket.c [hunk 3]
```diff
@@ -953,57 +955,113 @@ static int cache_and_validate_cert(nasp_socket *sc, char *key)
...
+		info.source_port = s->sk_num;
+		info.destination_port = port;
+	}
+
+	return info;
+}
+
+void add_sd_entry_tags_to_json(service_discovery_entry *sd_entry, JSON_Value *json)
+{
+	if (!json_object)
```
Comment (baluchicken) [inline socket.c:1010]:
```suggestion
	if (!json)
```

**Review by baluchicken — CHANGES_REQUESTED** [review]

File: device_driver.c (hunk 1 above)
Comment (waynz0r) [reply device_driver.c:398]: this gonna be fix in the logging related PR

File: socket.c (hunk 2 above)
Comment (waynz0r) [reply socket.c:936]: this gonna be fix in the logging related PR

File: sd.c [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+    }
+
+    u64 key = sd_entry_hash(entry->address, strlen(entry->address));
+
+    hash_add(table->htable, &entry->node, key);
+}
+
+static service_discovery_entry *sd_table_entry_get_locked(const char *address)
+{
+    struct service_discovery_entry *entry;
```
Comment (baluchicken) [inline sd.c:61]:
```suggestion
    service_discovery_entry *entry;
```

**Review by baluchicken — APPROVED** [review]

**Review by bonifaido — APPROVED** [review]

**Review by baluchicken — DISMISSED** [review]

**Review by baluchicken — APPROVED** [review]

**Review by bonifaido — APPROVED** [review]

**Pull request merged**